plugin/plugins/filesystem/nodes/base.go

```diff
@@ -0,0 +1,97 @@
+package fsnodes
+
+import (
+	"context"
+
+	"github.com/djdv/p9/p9"
+	"github.com/djdv/p9/unimplfs"
+	files "github.com/ipfs/go-ipfs-files"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+const ( //device - attempts to comply with standard multicodec table
+	dMemory = 0x2f
+	dIPFS   = 0xe4
+	dIPNS   = 0xe5
+	//TODO: decide what MFS/Files API should be; IPLD?
+	//dFiles = 0xe2
+)
+
+var _ p9.File = (*Base)(nil)
+
+// Base is a foundational file system node that provides common file metadata as well as stubs for unimplemented methods
+type Base struct {
+	// Provide stubs for unimplemented methods
+	unimplfs.NoopFile
+	p9.DefaultWalkGetAttr
+
+	// Storage for file's metadata
+	Qid      p9.QID
+	meta     p9.Attr
+	metaMask p9.AttrMask
+
+	// The base context should be derived from a long running context (like Background)
+	// and used to derive call specific contexts from
+	Ctx    context.Context
+	Logger logging.EventLogger
+
+	parent walkRef // parent should be set and used to handle ".." requests
+	child  walkRef // child is an optional field, used to hand off child requests to another file system
+}
+
+// IPFSBase is much like Base but extends it to hold IPFS specific metadata
+type IPFSBase struct {
+	Base
+
+	Path corepath.Resolved
+	core coreiface.CoreAPI
+
+	// you will typically want to derive a context from the base context within one operation (like Open)
+	// use it with the CoreAPI for something
+	// and cancel it in another operation (like Close)
+	// that pointer should be stored here between calls
+	cancel context.CancelFunc
+}
+
+func (ib *IPFSBase) Close() error {
+	if ib.cancel != nil {
+		ib.cancel()
```

Comment:
Given that this context is derived from the parent file's context, this will cascade the close all the way down through the children. If you open a parent, open a child relative to the parent, then close the parent, the child will be closed.

This could be causing the issues you've been noticing.

Reply:
This was partially the case. There were multiple issues around contexts, and inheritance.
Resolved in b6dee76e36d907f753e129d122bd8929b77da3d9, 3ab60290fd404e4f9d09628c973fe6a19c4134e1, and eaff0c93b4bf15113cdf02a519c584ae77aae89e

(thanks you for pointing this out when I was stumped)